CREATE DATABASE db_with_drop_failed;
-- Inject fault on content0 primary to error during prepare phase of 2PC.
SELECT gp_inject_fault2('end_prepare_two_phase', 'error', dbid, hostname, port)
FROM gp_segment_configuration WHERE content=0 and role='p';
 gp_inject_fault2 
------------------
 Success:
(1 row)

-- Should abort
DROP DATABASE db_with_drop_failed;
ERROR:  fault triggered, fault name:'end_prepare_two_phase' fault type:'error'  (seg0 127.0.0.1:25432 pid=38224)
-- The database should remain in catalog.
SELECT count(1) = 1 as result FROM pg_database WHERE datname='db_with_drop_failed';
 result 
--------
 t
(1 row)

-- Try creating a table in this database.  If it succeeds, we are sure
-- that the database directory and the catalog tables are still
-- present.
\! psql -d db_with_drop_failed -c "create table tbl_in_db_with_drop_failed()"
psql: FATAL:  database "db_with_drop_failed" does not exist
DETAIL:  The database subdirectory "base/16394" is missing.
-- Cleanup
SELECT gp_inject_fault2('end_prepare_two_phase', 'reset', dbid, hostname, port)
FROM gp_segment_configuration WHERE content=0 AND role='p';
 gp_inject_fault2 
------------------
 Success:
(1 row)

DROP DATABASE db_with_dropfailed;
ERROR:  database "db_with_dropfailed" does not exist
